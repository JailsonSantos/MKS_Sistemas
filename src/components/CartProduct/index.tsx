import { useState } from "react";
import { useDispatch } from "react-redux";
import { Minus, Plus, X } from "phosphor-react";
import { addTotal, deleteProduct, minusTotal } from "@/redux/cartRedux";
import {
  AreaButtonCart,
  AreaDescriptionPrice,
  AreaProduct,
  AreaQuantityProduct,
  CloseButtonCart,
  ProductButton,
  ProductQuantity,
  Container,
  ImageProduct,
  PriceProduct,
  TitleProduct
} from "./styles";

interface ProductProps {
  product: {
    id: number;
    name: string;
    description: string;
    photo: string;
    price: string;
    quantityOfProduct: number;
  }
}

export function CartProduct({ product }: ProductProps) {
  const dispatch = useDispatch();

  const [quantity, setQuantity] = useState(product.quantityOfProduct);

  const handleUpdateQuantityOfProduct = (type: string) => {
    if (type === "desc") {
      if (quantity > 1) {
        setQuantity(quantity - 1)

        const newTotal = 1 * Number(product.price)

        const newProduct = {
          id: product.id,
          name: product.name,
          description: product.description,
          photo: product.photo,
          price: product.price,
          quantityOfProduct: quantity - 1
        }

        dispatch(
          minusTotal(
            {
              ...newProduct,
              total: newTotal,
            }
          )
        );

      }
    } else {
      setQuantity(quantity + 1)

      const newTotal = 1 * Number(product.price)

      const newProduct = {
        id: product.id,
        name: product.name,
        description: product.description,
        photo: product.photo,
        price: product.price,
        quantityOfProduct: quantity + 1
      }

      dispatch(
        addTotal(
          {
            ...newProduct,
            total: newTotal,
          }
        )
      );
    }
  }

  const handleDeleteProductOfCart = () => {
    dispatch(
      deleteProduct(
        {
          ...product,
        }
      )
    );
  }

  let priceTotal = (quantity * Number(product.price)).toFixed(2);

  return (
    <Container>
      <AreaProduct>
        <ImageProduct src={product.photo} alt="Apple Watch" />
        <AreaDescriptionPrice>
          <TitleProduct>{product.name}</TitleProduct>
          <div>
            <AreaButtonCart>
              <span>Qtd:</span>
              <AreaQuantityProduct>
                <ProductButton onClick={() => handleUpdateQuantityOfProduct("desc")} >
                  <Minus size={16} weight="bold" />
                </ProductButton>

                <ProductQuantity>{product.quantityOfProduct}</ProductQuantity>

                <ProductButton onClick={() => handleUpdateQuantityOfProduct("inc")}>
                  <Plus size={16} weight="bold" />
                </ProductButton>
              </AreaQuantityProduct>
            </AreaButtonCart>
            <PriceProduct>RS {priceTotal}</PriceProduct>
          </div>
        </AreaDescriptionPrice>
      </AreaProduct>
      <CloseButtonCart onClick={handleDeleteProductOfCart}>
        <X />
      </CloseButtonCart>
    </Container>
  )
}
