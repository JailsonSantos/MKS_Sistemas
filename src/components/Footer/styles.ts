import styled from "styled-components";

export const Container = styled.div`
  bottom: 0;
  width: 100%;
  height: 2rem;
  display: flex;
  position: relative;
  align-items: center;
  justify-content: center;
  background: ${({ theme }) => theme.footer};

  span{
    font-size: 12px;
    font-weight: 400;
    line-height: 15px;
    color: ${({ theme }) => theme.textDark};
  }
`;