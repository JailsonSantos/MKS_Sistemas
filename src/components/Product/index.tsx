import { useDispatch } from "react-redux";
import { ShoppingBag } from "phosphor-react";
import Skeleton from 'react-loading-skeleton';
import { addProduct } from "@/redux/cartRedux";
import 'react-loading-skeleton/dist/skeleton.css';
import { Fragment, PropsWithChildren } from "react";

import {
  Container,
  AreaProduct,
  ImageProduct,
  PriceProduct,
  TitleProduct,
  FooterProduct,
  DescriptionProduct,
  AreaDescriptionPrice,
} from "./styles";

interface ProductProps {
  product: {
    id: number;
    name: string;
    brand: string;
    description: string;
    photo: string;
    price: string;
  },
  isLoading?: boolean;
}

import toast, { Toaster } from 'react-hot-toast';

export function Product({ product, isLoading }: ProductProps) {

  const dispatch = useDispatch();

  const notify = () => toast('Produto adicionado ao carrinho.');

  const handleAddNewProductToCart = () => {
    const priceFormatted = Number(product.price);
    const total = 1 * priceFormatted;

    const newProduct = {
      id: product.id,
      name: product.name,
      description: product.description,
      photo: product.photo,
      price: product.price,
      quantityOfProduct: 1
    }

    dispatch(
      addProduct(
        {
          ...newProduct,
          quantity: 1,
          total,
        }
      )
    );

    notify();
  };

  function InlineWrapperWithMargin({ children }: PropsWithChildren<unknown>) {
    return <span style={{ marginLeft: '0.2rem' }}>{children}</span>
  }

  return (
    <Container>
      <AreaProduct>
        {
          isLoading ?
            <Fragment>
              <Skeleton
                width="111px"
                height="138px"
              />

              <Skeleton
                width="100px"
                height="30px"
                borderRadius="0.5rem"
                wrapper={InlineWrapperWithMargin}
                count={2}
                inline
              />

              <Skeleton
                width="210px"
                height="50px"
                borderRadius="0.5rem"
              />
            </Fragment>
            :
            <Fragment>

              <ImageProduct src={product.photo} alt="Apple Watch" />

              <AreaDescriptionPrice>
                <TitleProduct>{product.name}</TitleProduct>
                <PriceProduct>RS {product.price}</PriceProduct>
              </AreaDescriptionPrice>

              <DescriptionProduct> {product.description}</DescriptionProduct>

            </Fragment>
        }
      </AreaProduct>

      {
        isLoading ?

          <Skeleton
            width="220px"
            height="35px"
            style={{
              borderTopLeftRadius: 0,
              borderTopRightRadius: 0,
              borderBottomLeftRadius: '0.5rem',
              borderBottomRightRadius: '0.5rem',
              position: 'relative',
              bottom: -10,
            }}
          />
          :
          <FooterProduct onClick={handleAddNewProductToCart}>
            <Toaster
              toastOptions={{
                style: {
                  background: "#28a745", color: '#FFFFFF',
                },
                duration: 1500
              }}
            />
            <ShoppingBag size={24} weight="bold" color="#FFFFFF" />
            COMPRAR
          </FooterProduct>
      }
    </Container>
  )
}