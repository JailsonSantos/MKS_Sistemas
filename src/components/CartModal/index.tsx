import { RootState } from '@/redux/store';
import { useSelector } from 'react-redux';
import { CartProduct } from '../CartProduct';
import { ShoppingCart, X } from "phosphor-react";
import { blackA, mauve } from '@radix-ui/colors';
import { styled, keyframes } from '@stitches/react';
import * as AlertDialogPrimitive from '@radix-ui/react-alert-dialog';
import {
  AreaTotalCart,
  BodyModal,
  ButtonCart,
  ConfirmPaymentCart,
  EmptyBodyCart,
  FooterCart,
  HeaderModal
} from './styles';

const overlayShow = keyframes({
  '0%': { opacity: 0 },
  '100%': { opacity: 1 },
});

const StyledOverlay = styled(AlertDialogPrimitive.Overlay, {
  backgroundColor: blackA.blackA9,
  position: 'fixed',
  inset: 0,
  '@media (prefers-reduced-motion: no-preference)': {
    animation: `${overlayShow} 150ms cubic-bezier(0.16, 1, 0.3, 1)`,
  },
});

const StyledContent = styled(AlertDialogPrimitive.Content, {
  backgroundColor: '#0F52BA',
  boxShadow: 'hsl(206 22% 7% / 35%) 0px 10px 38px -10px, hsl(206 22% 7% / 20%) 0px 10px 20px -15px',
  position: 'fixed',
  top: '0%',
  right: '0%',
  transform: 'translate(0%, 0%)',
  width: '85%',
  maxWidth: '486px',
  height: '100%',
  padding: '0',
  '&:focus': { outline: 'none' },
});

function Content({ children, ...props }: any) {
  return (
    <AlertDialogPrimitive.Portal>
      <StyledOverlay />
      <StyledContent {...props}>{children}</StyledContent>
    </AlertDialogPrimitive.Portal>
  );
}

const StyledTitle = styled(AlertDialogPrimitive.Title, {
  margin: 0,
  fontSize: 17,
  fontWeight: 700,
  color: '#ffffff',
});

const StyledDescription = styled(AlertDialogPrimitive.Description, {
  fontSize: 15,
  lineHeight: 1.5,
  marginBottom: 20,
  textAlign: 'center',
  color: mauve.mauve11,
});

// Exports
export const AlertDialogContent = Content;
export const AlertDialogTitle = StyledTitle;
export const AlertDialog = AlertDialogPrimitive.Root;
export const AlertDialogDescription = StyledDescription;
export const AlertDialogAction = AlertDialogPrimitive.Action;
export const AlertDialogCancel = AlertDialogPrimitive.Cancel;
export const AlertDialogTrigger = AlertDialogPrimitive.Trigger;

// Your app...
const Flex = styled('div', { display: 'flex' });

const Button = styled('button', {
  all: 'unset',
  display: 'inline-flex',
  alignItems: 'center',
  justifyContent: 'center',
  borderRadius: 4,
  padding: '0 5px',
  fontSize: 16,
  lineHeight: 1,
  fontWeight: 500,
  border: 'none',

  variants: {
    variant: {
      addProductInCart: {
        backgroundColor: 'transparent',
        color: '#005693',
        '&:hover': { cursor: 'pointer' },
        '&:focus': { outline: 'none' },
      },
      cancel: {
        backgroundColor: '#000000',
        border: '1px solid #000000',
        color: '#ffffff',
        width: '38px',
        height: '38px',
        borderRadius: '50%',
        padding: 0,
        '&:hover': {
          cursor: 'pointer',
          color: '#ffffff',
          border: '1px solid #ffffff',
          transition: 'all 0.2s ease',
        },
        '&:focus': { outline: 'none' },
      },
    },
  },
});

interface CartModalProps {
  quantity: number;
}

export function CartModal({ quantity }: CartModalProps) {

  const cart = useSelector((state: RootState) => state.cart);

  return (
    <AlertDialog>
      <AlertDialogTrigger asChild>

        <ButtonCart>
          <ShoppingCart size={24} weight="bold" />
          <span>{quantity}</span>
        </ButtonCart>

      </AlertDialogTrigger>
      <AlertDialogContent >

        <HeaderModal >
          <AlertDialogTitle css={
            {
              width: '100%',
              fontSize: '27px',
              fontWeight: '700',
              lineHeight: '32px',
            }
          }>
            Carrinho {<br />}
            de compras
          </AlertDialogTitle>
          <Flex css={{ cursor: 'pointer' }}>
            <AlertDialogCancel asChild>
              <Button variant="cancel">
                <X size={24} />
              </Button>
            </AlertDialogCancel>
          </Flex>
        </HeaderModal>

        <BodyModal>
          {
            cart.quantity > 0 ?
              cart?.products?.map(product => (
                <CartProduct key={product.id} product={product} />
              ))
              :
              <EmptyBodyCart>
                O carrinho est√° vazio, {<br />}
                Por favor Adicione algum produto.
              </EmptyBodyCart>
          }
        </BodyModal>

        <FooterCart>
          <AreaTotalCart>
            <span>
              Total
            </span>
            <span>
              R$ {cart.total.toFixed(2)}
            </span>
          </AreaTotalCart>

          <ConfirmPaymentCart total={cart.total}>
            Finalizar Compra
          </ConfirmPaymentCart>
        </FooterCart>
      </AlertDialogContent>
    </AlertDialog>
  );
}