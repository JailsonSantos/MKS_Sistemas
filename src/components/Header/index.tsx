
import { CartModal } from '../CartModal';
import { useSelector } from 'react-redux';
import { RootState } from '@/redux/store';
import { Container, Logo } from "./styles";
import * as Dialog from '@radix-ui/react-dialog';

export function Header() {

  const cart = useSelector((state: RootState) => state.cart);

  const quantity = cart?.quantity;

  return (
    <Container>
      <div>
        <Logo>
          MKS <span>Sistemas</span>
        </Logo>
      </div>

      <Dialog.Root>
        <CartModal quantity={quantity} />
      </Dialog.Root>

    </Container>
  )
}