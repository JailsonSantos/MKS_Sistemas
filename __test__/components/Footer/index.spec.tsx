import '@testing-library/jest-dom';
import { Footer } from '@/components/Footer';
import { render, screen } from '@testing-library/react';

describe('Footer component', () => {
  it("Shoud render component", () => {

    const { container } = render(<Footer />)

    expect(container).toMatchSnapshot()

  });

  it('Text exists in component', () => {

    render(<Footer />)

    expect(screen.getByText('MKS sistemas Â© Todos os direitos reservados.')).toBeInTheDocument();

  });
});