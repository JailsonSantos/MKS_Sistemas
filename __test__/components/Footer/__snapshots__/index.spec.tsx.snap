// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component Shoud render component 1`] = `
<div>
  <div
    class="sc-bcXHqe encDOS"
  >
    <span>
      MKS sistemas Â© Todos os direitos reservados.
    </span>
  </div>
</div>
`;
